### ISSUE TITLE
Keep alive failures

### ISSUE THREAD

**Issue by engebret** [Azure/azure-uamqp-python#135]
When trying to use keep_alive_interval on a set of servicebus queues (where the queues are configured, but no remote client is yet connected) I am hitting errors and unexpected state changes in the amqp connections.  Ultimately the connections fail and need to be recreated.  If remote clients are enabled AND message traffic is occurring this does not happen, so there must be some level of timeout causing this.

python 3.6.9
uamqp 1.2.5
azure-servicebus 0.50.2

Example python code snippet: 
self.input_queue = self.servicebus.get_queue(self.input_queue_name).get_receiver(keep_alive_interval = 30) 
self.output_queue = self.servicebus.get_queue(self.output_queue_name).get_sender(keep_alive_interval = 31) 

Resulting logs which show the AMQP state changes and failures attached below (note that there are two senders in this example).  The sends work one time then start failing.  Anyone have ideas on what might be occurring here?

DEBUG:uamqp.client:Keeping 'ReceiveClient' connection alive.
DEBUG:uamqp.client:Keeping 'SendClient' connection alive.
DEBUG:uamqp.client:Keeping 'SendClient' connection alive.
DEBUG:uamqp.client:Keeping 'ReceiveClient' connection alive.
DEBUG:uamqp.client:Keeping 'SendClient' connection alive.
INFO:uamqp.connection:Connection b'SBSender-5e63e211-e448-4a16-a869-206a95cb0d94' state changed from <ConnectionState.OPENED: 9> to <ConnectionState.DISCARDING: 12>
DEBUG:uamqp.client:Keeping 'SendClient' connection alive.
INFO:uamqp.connection:Connection b'SBSender-1831c47e-538e-4aa1-a9bb-6bb23253efa2' state changed from <ConnectionState.OPENED: 9> to <ConnectionState.DISCARDING: 12>
DEBUG:uamqp.client:Keeping 'ReceiveClient' connection alive.
DEBUG:uamqp.client:Keeping 'SendClient' connection alive.
INFO:uamqp.connection:Connection b'SBSender-5e63e211-e448-4a16-a869-206a95cb0d94' state changed from <ConnectionState.DISCARDING: 12> to <ConnectionState.DISCARDING: 12>
DEBUG:uamqp.client:Keeping 'SendClient' connection alive.
INFO:uamqp.connection:Connection b'SBSender-1831c47e-538e-4aa1-a9bb-6bb23253efa2' state changed from <ConnectionState.DISCARDING: 12> to <ConnectionState.DISCARDING: 12>
DEBUG:uamqp.client:Keeping 'ReceiveClient' connection alive.
DEBUG:uamqp.client:Keeping 'SendClient' connection alive.
INFO:uamqp.c_uamqp:b'Failure: sending socket failed 10053.' (b'D:\\a\\1\\s\\src\\vendor\\azure-uamqp-c\\deps\\azure-c-shared-utility\\adapters\\socketio_win32.c':b'socketio_send':422)
INFO:uamqp.c_uamqp:b'xio_send failed' (b'D:\\a\\1\\s\\src\\vendor\\azure-uamqp-c\\deps\\azure-c-shared-utility\\adapters\\tlsio_schannel.c':b'send_chunk':484)
INFO:uamqp.c_uamqp:b'send_chunk failed' (b'D:\\a\\1\\s\\src\\vendor\\azure-uamqp-c\\deps\\azure-c-shared-utility\\adapters\\tlsio_schannel.c':b'internal_send':525)
INFO:uamqp.c_uamqp:b'send failed' (b'D:\\a\\1\\s\\src\\vendor\\azure-uamqp-c\\deps\\azure-c-shared-utility\\adapters\\tlsio_schannel.c':b'tlsio_schannel_send':1263)
INFO:uamqp.c_uamqp:b'xio_send failed' (b'D:\\a\\1\\s\\src\\vendor\\azure-uamqp-c\\src\\saslclientio.c':b'saslclientio_send_async':1190)
INFO:uamqp.c_uamqp:b'Cannot send encoded bytes' (b'D:\\a\\1\\s\\src\\vendor\\azure-uamqp-c\\src\\connection.c':b'on_bytes_encoded':268)

**Comment by ramya-rao-a** [comment]
cc @KieranBrantnerMagee

**Comment by KieranBrantnerMagee** [comment]
Hey engebret;
Thanks for the logs and detail, and your patience thus far, we'll dig in and loop back here with any information or questions. 

(Appreciated for the ping and looping us in Ramya!)

**Comment by KieranBrantnerMagee** [comment]
To give a brief status update (@engebret you may have heard this response via email; realized I had an opportunity to unify it with this thread for this update so apologies if this is a repeat)

In investigation we tried a bunch of permutations of send and receive but were having trouble reproducing the failure as I understood it.  Please see [this gist](https://gist.github.com/KieranBrantnerMagee/2daa00264186531180ff91c0159152fc) for how we were attempting to repro; attempted with various permutations and delays but all results were as expected.

Some additional detail on how send and receive are being called might be useful even insofar as "the above sample looks similar", as would any occurrence "symptoms", e.g. how often does it occur, is it reliable or sporadic, after N minutes/messages, etc.  You mention fails after one send, but I wanted to be precise for my own understanding on the full context, and the extent of what "fails" mean, does sending/receipt resume or halt forever?

As a final note, I'd mention that the xio_send failed error can occur due to a broad range of symptoms including transient interruption; this may not be the case here, I bring it up only because it was a red herring in a recent investigation.  Being able to repro reliably would be the biggest step towards a root cause.

Thanks as always for the patience and your time.

**Comment by KieranBrantnerMagee** [comment]
Hey engebret; 
given the latest report that the prospective fix didn't seem to do the trick, we're digging back in with priority; sincere apologies for the issue.  Echoing the updates here that I sent via email, per your mention about eyeballing this thread for details (we're more than happy to use this as the primary channel)

1.	Since it was mentioned that closing and reopening the sender mitigates the issue, is having a try/except around the send() and calling “sender.reconnect()” prohibitive/not do the trick?  There are scenarios where the link may be killed (e.g. by the service after 10 minutes idle) after which calling send() may fail due to the receiver being in a errored state and requiring a reconnect(), which sounds _suspiciously_ like what is being mentioned.

2.	The test case we were provided never actually formally opens or closes the sender in question.  I won’t make any assertions that this makes your issue go away, but it’s something to check off, and can be easily added with a “with servicebus.get_queue(output_queue_name).get_sender() as output_queue:” stanza, or by calling open() and close() manually at the start and end of the sender lifespan.


    a.	While I’m unfortunately still unable to repro your failure even with the sample, (It simply continues to keep-alive until the link is force-shutdown with a very distinct error) I can handwave at cases where the above combined with a non-halted keep-alive might manifest strange behavior.

3.	The questions we posed upthread would still be useful, namely any more reproducible data around how to generate the failure, its occurrence rate, and whether the samples we’ve been trying are indeed representative.   

4.	In lieu of that, the verbosity of logging can be turned up if the ServiceBusClient is initialized with “debug=True”; this may provide additional insights by allowing us to see the full network traces.

Thanks for all your patience thus far.

**Comment by engebret** [comment]
I am able to recreate the issue with the simple test script so would like to keep the focus on that example & try to understand why you see different behavior.  That implies a potential environment difference or perhaps configuration of the server side is different in some way.

One thing I'd like to understand better are the default timeout settings in the stack.  Should a keep_alive_interval = 30 be sufficient to avoid hitting any of the default timeouts?  Is there a good reference for various timeouts & default values which might be in play for this example?

**Comment by engebret** [comment]
Here is some debug output from running the simple test application with keep alive = 30s.  Note that I have some extra debugging code in the uamqp/connection.py code.  The main concern I have here is the unexpected state transition from that code:
https://github.com/Azure/azure-uamqp-python/blob/master/uamqp/connection.py#L178

<...>
DEBUG:uamqp.client:Message sent: <MessageSendResult.Ok: 0>, []
*** hit <enter> to continue 1
DEBUG:uamqp.client:Keeping 'SendClient' connection alive.
DEBUG:uamqp.client:Keeping 'SendClient' connection alive.
DEBUG:uamqp.client:Keeping 'SendClient' connection alive.

**_Output at ~2:00 minutes:_**
DEBUG:uamqp.client:Keeping 'SendClient' connection alive.
INFO:uamqp.connection:Connection b'SBSender-9a656aba-8ad0-4dfc-909b-4bae641602d2' state changed from <ConnectionState.OPENED: 9> to <ConnectionState.DISCARDING: 12>

**_Output at ~2:30 minutes:_**
DEBUG:uamqp.client:Keeping 'SendClient' connection alive.
INFO:uamqp.connection:Connection b'SBSender-9a656aba-8ad0-4dfc-909b-4bae641602d2' state changed from <ConnectionState.DISCARDING: 12> to <ConnectionState.DISCARDING: 12>

**_Output at ~3:00 minutes:_** 
DEBUG:uamqp.client:Keeping 'SendClient' connection alive.
INFO:uamqp.c_uamqp:b'Failure: sending socket failed 10053.' (b'D:\\a\\1\\s\\src\\vendor\\azure-uamqp-c\\deps\\azure-c-sh
ared-utility\\adapters\\socketio_win32.c':b'socketio_send':422)
INFO:uamqp.c_uamqp:b'xio_send failed' (b'D:\\a\\1\\s\\src\\vendor\\azure-uamqp-c\\deps\\azure-c-shared-utility\\adapters
\\tlsio_schannel.c':b'send_chunk':484)
INFO:uamqp.c_uamqp:b'send_chunk failed' (b'D:\\a\\1\\s\\src\\vendor\\azure-uamqp-c\\deps\\azure-c-shared-utility\\adapte
rs\\tlsio_schannel.c':b'internal_send':525)
INFO:uamqp.c_uamqp:b'send failed' (b'D:\\a\\1\\s\\src\\vendor\\azure-uamqp-c\\deps\\azure-c-shared-utility\\adapters\\tl
sio_schannel.c':b'tlsio_schannel_send':1263)
INFO:uamqp.c_uamqp:b'xio_send failed' (b'D:\\a\\1\\s\\src\\vendor\\azure-uamqp-c\\src\\saslclientio.c':b'saslclientio_se
nd_async':1190)
INFO:uamqp.c_uamqp:b'Cannot send encoded bytes' (b'D:\\a\\1\\s\\src\\vendor\\azure-uamqp-c\\src\\connection.c':b'on_byte
s_encoded':268)
INFO:uamqp.connection:Connection b'SBSender-9a656aba-8ad0-4dfc-909b-4bae641602d2' state changed from <ConnectionState.DISCARDING: 12> to <ConnectionState.END: 13>

INFO:uamqp.connection:Connection with ID b'SBSender-9a656aba-8ad0-4dfc-909b-4bae641602d2' unexpectedly in an error state
. Closing: False, Error: None
ConnectionState.END
ConnectionState.DISCARDING
INFO:uamqp.c_uamqp:CBS error occured on connection b'SBSender-9a656aba-8ad0-4dfc-909b-4bae641602d2'.
INFO:uamqp.sender:Message sender b'sender-link-3b9e25da-3354-40b0-b2d5-aa588131f629' state changed from <MessageSenderState.Open: 2> to <MessageSenderState.Idle: 0> on connection: b'SBSender-9a656aba-8ad0-4dfc-909b-4bae641602d2'
INFO:uamqp.connection:Connection b'SBSender-9a656aba-8ad0-4dfc-909b-4bae641602d2' state changed from <ConnectionState.END: 13> to <ConnectionState.DISCARDING: 12>
DEBUG:uamqp.client:Keeping 'SendClient' connection alive.
WARNING:uamqp.connection:ConnectionClose('ErrorCodes.UnknownError: Connection in an unexpected error state.',)
INFO:uamqp.client:Connection keep-alive for 'SendClient' failed: ConnectionClose('ErrorCodes.UnknownError: Connection in an unexpected error state.',).

**Comment by engebret** [comment]
Experimenting with a keep_alive_interval = 5 appears to remove the issue of an illegal state transition exception being raised.  Eventually the link detatches after we hit a 10 minute IdleTimerExpired (messages sent after this timeout raise an exception & are retried), but I suspect that is an expected behavior ?

**Comment by KieranBrantnerMagee** [comment]
So I _may_ have been able to reproduce a similar failure, but I'm still nervous as it's not 100% in parallel with what you've posted, (e.g. doesn't go away with keep_alive_interval=5, only when entirely removing keepalive from the sender, and manifests as a harder crash than you've seen) and at least in my case seems to come down to providing keep_alive_interval to the sender at all; if I remove it the behavior is as expected. 

My failure looks like the following, if it rings any bells:
Fatal Python error: GC object already tracked

Current thread 0x00008238 (most recent call first):
  File "c:\program files (x86)\microsoft visual studio\2019\community\common7\ide\extensions\microsoft\python\core\Packages\ptvsd\_vendored\pydevd\pydevd_file_utils.py", line 332 in _NormPaths
  File "c:\program files (x86)\microsoft visual studio\2019\community\common7\ide\extensions\microsoft\python\core\Packages\ptvsd\_vendored\pydevd\pydevd_file_utils.py", line 698 in get_abs_path_real_path_and_base_from_file
  File "C:\Program Files\Python36\lib\threading.py", line 911 in _bootstrap_inner
  File "C:\Program Files\Python36\lib\threading.py", line 884 in _bootstrap
  File "c:\program files (x86)\microsoft visual studio\2019\community\common7\ide\extensions\microsoft\python\core\Packages\ptvsd\_vendored\pydevd\_pydev_bundle\pydev_monkey.py", line 667 in __call__

Re: your question about expected timeouts, Unfortunately I don't believe there's such a doc, but I've put a task on us to produce one.  In leiu of that however, I'd call out the lock/session expiries set on queue creation, and the 10 minute service-side timeout.  (it is as you suggest an expected behavior, the service kills the SB link after 10 minutes idle as a defensive measure, we have mitigations to allow disabling this pending our next release if it becomes relevent to you; this was the initial fix we had sent under suspicion it might be what you had been seeing.)

That said, I would say that I candidly don't expect keep_alive_interval to even be regularly necessary on the sender:  since it's basically not stateful, if a non-retriable error occurs, I'd typically just catch that and try a reconnect/resend.  May I ask what the specific need for the sender keepalive was as opposed to trying a reconnect?  (If the answer is "save the boilerplate" that's totally reasonable, just wanted to understand if that was a feasible workaround)

Your focus on version parity is well taken, candidly I've been running a slightly different 3.6 revision because the later 3.6.* revisions don't have installers for windows and would require building from source, and I (perhaps falsely) assumed that I couldn't imagine how that would cause the delta we're seeing; it does beg the question if there's an underlying OS delta (I am on win10) as well that might be related.  I can work to set up a more precisely comparable env if we think that's a promising avenue.

All that said, per the failure above there's at least something we can dig into on our end, but let me know if this echoes what you're seeing or if the proposed workarounds (remove keepalive from sender, use a reconnect/retry pattern on disconnect) would be of assistance.

**Comment by engebret** [comment]
I have not seen the GC fatal error you report so that appears to be something new.

The only reason I was using the keep alive was to try to avoid the problem we were seeing where after just a couple minutes (or less) of no sent messages the exception I reported was thrown.  My thought was it would keep the connection in a good state if we are not sending actual data messages.  The short (5s) keep alive does appear to help, so there is a potential clue.  Maybe it is OS related -- our dev system is a pretty old version of Windows server.

**Comment by yunhaoling** [comment]
Hi @engebret , thanks for providing the additional context. I did some investigation into the C library (azure-uamqp-c which is used by azure-uamqp-python) and found that the connection state would be set to `CONNECTION_STATE_DISCARDING` when the connection is closed with error.

There're a certain amount of cases when connection would be closed with error:
- fail to parse bytes received from the socket
- receive illegal amqp performative negotiation, e.g. same connection being opened twice
- etc.

As you may know, the underlying implementation of uamqp.Connection is socket.
From your description, I think the problem here might be with the socket, the socket might go into some kind of "inactive" status thus leading to the uamqp library getting error-ed/incomplete bytes from the socket -- and connection finally crashes.

With that being said, can you turn on the amqp logging tracing when you run the code? It would be pretty straightforward:
```
# pass debug=True into from_connection_string or constructor
servicebus = ServiceBusClient.from_connection_string(endpoint, debug=True)
```

This would give more detailed info about the amqp connection activity which might be helpful for us to locate the issue.

**Comment by engebret** [comment]
Output below captured with debug=True & is annotated with approximate timestamps.  I think the theory about an issue with the underlying socket is a good possibility.  It would explain why my behavior is different than what you have been able to recreate (we are on an older version of Windows Server).  It does seem to me though that the keep alive messages should essentially help avoid timeout based closures; how are those keep alive messages different at a socket layer than the normal messages we are sending?  When 'normal' messages are sent on a regular basis we do not see this issue.  That is why I started to experiment with keep alive messages with the expectation being any socket layer timeout issues would be avoided.


## <... previous messages sent ...>

*** Sending msg 9
DEBUG:uamqp.c_uamqp:Deallocating cProperties
DEBUG:uamqp.c_uamqp:Destroying cProperties
DEBUG:uamqp.c_uamqp:Deallocating cHeader
DEBUG:uamqp.c_uamqp:Destroying cHeader
INFO:uamqp.c_uamqp:b'Header:* {}Properties:* {}-> [TRANSFER]* {2,9,<09 00 00 00>,0,false,false}'
DEBUG:uamqp.c_uamqp:Deallocating cMessage
DEBUG:uamqp.c_uamqp:Destroying cMessage
INFO:uamqp.c_uamqp:b'<- [DISPOSITION]* {true,9,NULL,true,* {},NULL}'
DEBUG:uamqp.c_uamqp:Wrapping value type: <AMQPType.ListValue: 18>
DEBUG:uamqp.c_uamqp:Deallocating 'ListValue'
DEBUG:uamqp.c_uamqp:Destroying 'ListValue'
DEBUG:uamqp.client:Message sent: <MessageSendResult.Ok: 0>, []
*** Sending msg 10
DEBUG:uamqp.c_uamqp:Deallocating cProperties
DEBUG:uamqp.c_uamqp:Destroying cProperties
DEBUG:uamqp.c_uamqp:Deallocating cHeader
DEBUG:uamqp.c_uamqp:Destroying cHeader
INFO:uamqp.c_uamqp:b'Header:* {}Properties:* {}-> [TRANSFER]* {2,10,<0A 00 00 00>,0,false,false}'
DEBUG:uamqp.c_uamqp:Deallocating cMessage
DEBUG:uamqp.c_uamqp:Destroying cMessage
INFO:uamqp.c_uamqp:b'<- [DISPOSITION]* {true,10,NULL,true,* {},NULL}'
DEBUG:uamqp.c_uamqp:Wrapping value type: <AMQPType.ListValue: 18>
DEBUG:uamqp.c_uamqp:Deallocating 'ListValue'
DEBUG:uamqp.c_uamqp:Destroying 'ListValue'
DEBUG:uamqp.client:Message sent: <MessageSendResult.Ok: 0>, []

## At this point the script stops sending messages & is waiting for input to continue.
### Keep alive messages continue.

*** hit <enter> to continue 1

## 20s after last message sent: 
DEBUG:uamqp.client:Keeping 'SendClient' connection alive.

## 50s after last message sent (30s keep alive time out):
DEBUG:uamqp.client:Keeping 'SendClient' connection alive.
**INFO:uamqp.c_uamqp:b'<- Empty frame'**

## 1:20s after last message sent :
DEBUG:uamqp.client:Keeping 'SendClient' connection alive.

## 1:50 after last message sent:
DEBUG:uamqp.client:Keeping 'SendClient' connection alive.
INFO:uamqp.c_uamqp:b'-> [CLOSE]* {* {amqp:internal-error,No frame received for the idle timeout}}'
INFO:uamqp.connection:Connection b'SBSender-ca374467-6426-4507-89ad-63784071efb0' **state changed from <ConnectionState.OPENED: 9> to <ConnectionState.DISCARDING: 12>**

## 2:20 after last message sent:
DEBUG:uamqp.client:Keeping 'SendClient' connection alive.
INFO:uamqp.c_uamqp:b'-> [CLOSE]* {* {amqp:internal-error,No frame received for the idle timeout}}'
INFO:uamqp.connection:Connection b'SBSender-ca374467-6426-4507-89ad-63784071efb0' state changed from <ConnectionState.DISCARDING: 12> to <ConnectionState.DISCARDING: 12>

## 2:50 after last message 
DEBUG:uamqp.client:Keeping 'SendClient' connection alive.
INFO:uamqp.c_uamqp:b'Failure: sending socket failed 10053.' (b'D:\\a\\1\\s\\src\\vendor\\azure-uamqp-c\\deps\\azure-c-shared-utility\\adapters\\socketio_win32.c':b'socketio_send':422)
INFO:uamqp.c_uamqp:b'xio_send failed' (b'D:\\a\\1\\s\\src\\vendor\\azure-uamqp-c\\deps\\azure-c-shared-utility\\adapters\\tlsio_schannel.c':b'send_chunk':484)
INFO:uamqp.c_uamqp:b'send_chunk failed' (b'D:\\a\\1\\s\\src\\vendor\\azure-uamqp-c\\deps\\azure-c-shared-utility\\adapters\\tlsio_schannel.c':b'internal_send':525)
INFO:uamqp.c_uamqp:b'send failed' (b'D:\\a\\1\\s\\src\\vendor\\azure-uamqp-c\\deps\\azure-c-shared-utility\\adapters\\tlsio_schannel.c':b'tlsio_schannel_send':1263)
INFO:uamqp.c_uamqp:b'xio_send failed' (b'D:\\a\\1\\s\\src\\vendor\\azure-uamqpc\\src\\saslclientio.c':b'saslclientio_send_async':1190)
INFO:uamqp.c_uamqp:b'Cannot send encoded bytes' (b'D:\\a\\1\\s\\src\\vendor\\azure-uamqp-c\\src\\connection.c':b'on_bytes_encoded':268)
INFO:uamqp.connection:Connection b'SBSender-ca374467-6426-4507-89ad-63784071efb0' state changed from <ConnectionState.DI
SCARDING: 12> to <ConnectionState.END: 13>
**INFO:uamqp.connection:Connection with ID b'SBSender-ca374467-6426-4507-89ad-63784071efb0' unexpectedly in an error state. Closing: False, Error: None
ConnectionState.END
ConnectionState.DISCARDING**
INFO:uamqp.c_uamqp:CBS error occured on connection b'SBSender-ca374467-6426-4507-89ad-63784071efb0'.
INFO:uamqp.sender:Message sender b'sender-link-297f69a5-394a-4863-8732-5149abbe5bdc' state changed from <MessageSenderState.Open: 2> to <MessageSenderState.Idle: 0> on connection: b'SBSender-ca374467-6426-4507-89ad-63784071efb0'
INFO:uamqp.c_uamqp:b'-> [CLOSE]* {* {amqp:internal-error,No frame received for the idle timeout}}'
INFO:uamqp.connection:Connection b'SBSender-ca374467-6426-4507-89ad-63784071efb0' state changed from <ConnectionState.END: 13> to <ConnectionState.DISCARDING: 12>

## 3:20 after last message sent: 
DEBUG:uamqp.client:Keeping 'SendClient' connection alive.
WARNING:uamqp.connection:ConnectionClose('ErrorCodes.UnknownError: Connection in an unexpected error state.',)
INFO:uamqp.client:Connection keep-alive for 'SendClient' failed: ConnectionClose('ErrorCodes.UnknownError: Connection in an unexpected error state.',).

**Comment by yunhaoling** [comment]
@engebret, thanks for providing the log, it definitely helps a lot!

I noticed the weird part in your latest post:
```
INFO:uamqp.c_uamqp:b'-> [CLOSE]* {* {amqp:internal-error,No frame received for the idle timeout}}'
```

I've searched the code of the C library, the only place that it would log such error is that when a local connection idle-timeout is set and that idle-timeout expires:

https://github.com/Azure/azure-uamqp-python/blob/803810434d5696b42641abfea9074069be446c29/src/vendor/azure-uamqp-c/src/connection.c#L1775-L1791

As far as I know, from ServiceBus API surface it can't be set which means the code should not hit local connection idle timeout expiration.

Have you tweaked the ServiceBus code such that a `uamp.SenderClient(Async)` is created with ["idle_timeout"](https://github.com/Azure/azure-uamqp-python/blob/803810434d5696b42641abfea9074069be446c29/uamqp/client.py#L453-L455) passed?

--------------------------------------------------------------------------------------------------

I would try to explain the underlying mechanism which might gave both of us some inspiration. (not taking local idle timeout into account.)

1. The remote idle timeout is 240s -- it's set by the service side.
2. Every keep_alive_interval (in your case is 30s) the client side would
    1) check whether time has passed 120s [(idle_timeout_empty_frame_send_ratio*remote_idle_timeout)](https://github.com/Azure/azure-uamqp-python/blob/803810434d5696b42641abfea9074069be446c29/src/vendor/azure-uamqp-c/src/connection.c#L878) since last frame sent
    2) if idle timeout> 120s, then try to [send empty frame (which is a special message type) out](https://github.com/Azure/azure-uamqp-python/blob/803810434d5696b42641abfea9074069be446c29/src/vendor/azure-uamqp-c/src/connection.c#L1793-L1827).
3. 10 mins of uamqp.Link level idle timeout hits, this can't be avoided if in 10 mins there's no real sending to a specific ServiceBus entity -- uamqp.Link is the layer where data transfer inside a ServiceBus entity happens and it's setting of the Service Side.

------------------------------------------------------------------------

My loggings:

*** Sending msg 1
2020-03-30 17:19:48,622 Deallocating cProperties
2020-03-30 17:19:48,623 Destroying cProperties
2020-03-30 17:19:48,624 Deallocating cHeader
2020-03-30 17:19:48,624 Destroying cHeader
2020-03-30 17:19:48,625 b'Header:* {}Properties:* {}-> [TRANSFER]* {2,10,<0A 00 00 00>,0,false,false}'
2020-03-30 17:19:48,625 Deallocating cMessage
2020-03-30 17:19:48,626 Destroying cMessage
2020-03-30 17:19:48,643 b'<- [DISPOSITION]* {true,10,NULL,true,* {},NULL}'
2020-03-30 17:19:48,643 Wrapping value type: <AMQPType.ListValue: 18>
2020-03-30 17:19:48,645 Deallocating 'ListValue'
2020-03-30 17:19:48,646 Destroying 'ListValue'
2020-03-30 17:19:48,647 Message sent: <MessageSendResult.Ok: 0>, []
*** hit <enter> to continue 1
2020-03-30 17:20:09,576 Keeping 'SendClient' connection alive.
2020-03-30 17:20:40,598 Keeping 'SendClient' connection alive.
2020-03-30 17:21:11,620 Keeping 'SendClient' connection alive.
2020-03-30 17:21:42,640 Keeping 'SendClient' connection alive.
2020-03-30 17:22:13,664 Keeping 'SendClient' connection alive.
2020-03-30 17:22:13,664 b'-> Empty frame'
2020-03-30 17:22:44,688 Keeping 'SendClient' connection alive.
2020-03-30 17:22:44,689 b'<- [FLOW]* {11,5000,2,65535,NULL,NULL,NULL,NULL,NULL,true,NULL}'
2020-03-30 17:23:15,718 Keeping 'SendClient' connection alive.
2020-03-30 17:23:46,740 Keeping 'SendClient' connection alive.
2020-03-30 17:24:17,760 Keeping 'SendClient' connection alive.
2020-03-30 17:24:17,760 b'-> Empty frame'
2020-03-30 17:24:48,785 Keeping 'SendClient' connection alive.
2020-03-30 17:25:19,808 Keeping 'SendClient' connection alive.
2020-03-30 17:25:50,832 Keeping 'SendClient' connection alive.
2020-03-30 17:25:50,832 b'<- [FLOW]* {11,5000,2,65535,NULL,NULL,NULL,NULL,NULL,true,NULL}'
2020-03-30 17:26:21,857 Keeping 'SendClient' connection alive.
2020-03-30 17:26:21,858 b'-> Empty frame'
2020-03-30 17:26:52,881 Keeping 'SendClient' connection alive.
2020-03-30 17:27:23,911 Keeping 'SendClient' connection alive.
2020-03-30 17:27:54,936 Keeping 'SendClient' connection alive.
2020-03-30 17:28:25,956 Keeping 'SendClient' connection alive.
2020-03-30 17:28:25,956 b'-> Empty frame'
2020-03-30 17:28:56,978 Keeping 'SendClient' connection alive.
2020-03-30 17:28:56,978 b'<- [FLOW]* {11,5000,2,65535,NULL,NULL,NULL,NULL,NULL,true,NULL}'
2020-03-30 17:29:28,001 Keeping 'SendClient' connection alive.
2020-03-30 17:29:59,022 Keeping 'SendClient' connection alive.
2020-03-30 17:29:59,022 b"<- [DETACH]* {2,true,* {amqp:link:detach-forced,The link 'G62:320228925:sender-link-50d109d3-1b7a-40cd-babe-9bc52869b49c' is force detached. Code: publisher(li
nk2552469). Details: AmqpMessagePublisher.IdleTimerExpired: Idle timeout: 00:10:00.,NULL}}"
2020-03-30 17:29:59,024 b'-> [DETACH]* {2,true}'
2020-03-30 17:29:59,025 Received Link detach event: b'amqp:link:detach-forced'
Link: b'sender-link-50d109d3-1b7a-40cd-babe-9bc52869b49c'
Description: b"The link 'G62:320228925:sender-link-50d109d3-1b7a-40cd-babe-9bc52869b49c' is force detached. Code: publisher(link2552469). Details: AmqpMessagePublisher.IdleTimerExpired:
 Idle timeout: 00:10:00."
Details: None
Retryable: False
Connection: b'SBSender-f1e210bf-38c0-40b8-ac95-cb84d0899662'
2020-03-30 17:29:59,026 Deallocating cError
2020-03-30 17:29:59,027 Message sender b'sender-link-50d109d3-1b7a-40cd-babe-9bc52869b49c' state changed from <MessageSenderState.Open: 2> to <MessageSenderState.Error: 4> on connection
: b'SBSender-f1e210bf-38c0-40b8-ac95-cb84d0899662'

**Comment by engebret** [comment]
Thank you for the additional information & ideas.  I went back through the code I was instrumenting to debug & found a spot I had missed when reverting to the original code.  My output now matches what you have (for the simple test), so I now think that the recent fix that was provided did in fact address the issue we had been seeing where a timeout less than 10m was causing exceptions on a message send.

Next I would like to better understand the uamqp.Link level idle timeout (10 min).   It sounds like that is a server side setting which we cannot adjust, and keep alive messages will not avoid, correct?  Is there an example of best practices to handle & ensure messages are not dropped?  The implementation that I have appears to be handling the exception properly but I would like to make sure there is nothing missed.

Here is the output from my traces through one of these timeouts:
2020-03-31 12:58:37,941 - algo - DEBUG - Messaging manager send_response try 1
2020-03-31 12:58:38,668 - algo - DEBUG - Messaging manager send_response done
2020-03-31 12:58:38,669 - algo - DEBUG - Messaging manager send_response msg.complete start
2020-03-31 12:58:38,669 - algo - DEBUG - Messaging manager send_response msg.complete done
2020-03-31 12:58:38,669 - algo - WARNING - response sent for message b'97d2e2922e18403e8b3ec7dc17cd5d0c' - 1445.4 ms

2020-03-31 13:09:24,138 - algo - DEBUG - received message b'8c0d3f1b7f5f45e89d2a65bfaa5fe283'. total received 31
2020-03-31 13:09:24,920 - algo - DEBUG - Messaging manager send_response try 1
2020-03-31 13:09:24,921 - algo - ERROR - Messaging manager error sending response on try 1
Traceback (most recent call last):
  File "D:\dev\python-install\python\envs\datascience\lib\site-packages\azure\servicebus\send_handler.py", line 105, in
send
    self._handler.send_message(message.message)
  File "D:\dev\python-install\python\envs\datascience\lib\site-packages\uamqp\client.py", line 726, in send_message
    running = self.do_work()
  File "D:\dev\python-install\python\envs\datascience\lib\site-packages\uamqp\client.py", line 400, in do_work
    if not self.client_ready():
  File "D:\dev\python-install\python\envs\datascience\lib\site-packages\uamqp\client.py", line 384, in client_ready
    if not self._client_ready():
  File "D:\dev\python-install\python\envs\datascience\lib\site-packages\uamqp\client.py", line 532, in _client_ready
    if self.message_handler.get_state() == constants.MessageSenderState.Error:
  File "D:\dev\python-install\python\envs\datascience\lib\site-packages\uamqp\sender.py", line 185, in get_state
    raise self._error
uamqp.errors.LinkDetach: ErrorCodes.LinkDetachForced: The link 'G24:5339991:sender-link-90ea27cf-2a35-4a00-9ad3-accf1448
434c' is force detached. Code: publisher(link3218193). Details: AmqpMessagePublisher.IdleTimerExpired: Idle timeout: 00:
10:00.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\Users\dengebretsen\git5\AlgorithmServer\Alg\ApplicationPackageRoot\Pkg\Code\messaging.
py", line 179, in send_response
    self.output_queue.send(response)
  File "D:\dev\python-install\python\envs\datascience\lib\site-packages\azure\servicebus\send_handler.py", line 107, in send
    raise MessageSendFailed(e) azure.servicebus.common.errors.MessageSendFailed: Message failed to send. Error: ErrorCodes.LinkDetachForced: The link ' G24:5339991:sender-link-90ea27cf-2a35-4a00-9ad3-accf1448434c' is force detached. Code: publisher(link3218193). Details: AmqpMessagePublisher.IdleTimerExpired: Idle timeout: 00:10:00.

2020-03-31 13:09:26,398 - algo - DEBUG - Messaging manager send_response try 2
2020-03-31 13:09:27,112 - algo - DEBUG - Messaging manager send_response done
2020-03-31 13:09:27,112 - algo - DEBUG - Messaging manager send_response msg.complete start
2020-03-31 13:09:27,113 - algo - DEBUG - Messaging manager send_response msg.complete done
2020-03-31 13:09:27,113 - algo - WARNING - response sent for message b'8c0d3f1b7f5f45e89d2a65bfaa5fe283' - 2975.3

Send code & exception handling:
            while not msg.expired and not sent:
                try:
                    send_counter += 1
                    # Additional debug logging here for a case where the send potentially blocked.
                    self.logger.debug('Messaging manager send_response try {}'.format(send_counter))
                    self.output_queue.send(response)
                    self.logger.debug('Messaging manager send_response done')
                    sent = True
                except ServiceBusError as e:
                    self.logger.exception('Messaging manager error sending response on try {}'.format(send_counter), extra=msg.user_properties)
                    self.output_queue.close()
                    self.output_queue = self.servicebus.get_queue(self.output_queue_name).get_sender(keep_alive_interval = 30) 
                    self.output_queue.open()
                # Bubble OS exceptions up as it is not clear any action we take to retry will help.
                except OSError as eOS:
                    self.logger.exception('OS Exception: error sending response on try {}'.format(send_counter), extra=msg.user_properties)
                    raise

**Comment by yunhaoling** [comment]
@engebret sorry about not getting to you sooner.

Yes, your understanding is correct.

I think you're doing good, generally retry practice sounds good to me.

There is a [`reconnect()`](https://github.com/Azure/azure-sdk-for-python/blob/300d00e56a80711230902fcd285e347ac0de121b/sdk/servicebus/azure-servicebus/azure/servicebus/send_handler.py#L195-L209) method on the `Sender` which should properly handle shutting down old uamqp objects and create new uamqp handler objects.

```python
with sender:
    msg = Message("body")
    try:
        sender.send(msg)
    except Exception:
        sender.reconnect()
        sender.send(msg)
```


Apart from that, we'll address the issue in our next SB release which would do internal retry. We don’t have a canonical ETA yet since we want to get some additional bugfixes in but I’d certainly hope to have news within the next few weeks.

**Comment by engebret** [comment]
@yunhaoling as your team sorts out the data for that next release please get the date to me as soon as you can.  We would like to roll this update into our next product update which has a timeline to wrap up development and start test by the start of May.

Thanks for the help & updates!

**Comment by engebret** [comment]
Any update on availability of the next release ?

**Comment by KieranBrantnerMagee** [comment]
Hey Engebret,
needed to sync with the powers that be to get a nod for this, but we're looking to have a release out within the next 2-3 weeks; (giving a pessimistic time window just in case). will loop back as I get an exact date.

**Comment by yunhaoling** [comment]
Hi @engebret , apologies for the delayed response.

And apologies one more time as it's completely my fault that I didn't take a careful look into the `reconnect` inside a Sender -- `reconnect` would internally send out pending messages.

I've prepared a patch for you which integrates the `send` with `reconnect` internally, can you try this patch to see if the fix solves your problem? -- If it works, we would get the fix included in our coming formal release.

Patch file: [servicebus-retry-send-patch.zip](https://github.com/Azure/azure-uamqp-python/files/4578497/servicebus-retry-send-patch.zip), you just need to unzip the zip file and `pip install azure_servicebus-0.50.3b1-py2.py3-none-any.whl` (whl file inside the zip file)

(And if you're interested in implementation, my PR for the fix is here: https://github.com/Azure/azure-sdk-for-python/pull/11203)

**Comment by engebret** [comment]
Hi @yunhaoling, I've been testing this change today, and believe I have found a case missing in your patch that covers the actual error I'm hitting (LinkDetatch).  For the two code blocks where you have:

        except (errors.ConnectionClose,
                errors.AuthenticationException,
                errors.MessageHandlerError):

The case of 'errors.LinkDetatch' is probably also required.   I'm still debugging, but preliminary results indicate this will be needed.  Please let me know what you think.

**Comment by yunhaoling** [comment]
Hi @engebret , thanks for pointing that out, I think you're right -- I've checked error handling and found out that LinkDetach indeed requires reconnection.

with that being said, I built another patch whl file with that being addressed, would you be able to test that one more time?

[servicebus-retry-send-patch-0505.zip](https://github.com/Azure/azure-uamqp-python/files/4585108/servicebus-retry-send-patch-0505.zip)

Appreciate your collaboration and patience on this very much. Let me know whether it works for you.

**Comment by engebret** [comment]
Hi @yunhaoling - this most recent patch is working better for my case, thank you!  The LinkDetatch after 10m is now handled transparently which improves the client code complexity.

Are there any updates on the timeline to get the various updates out into an official release?

Thanks

**Comment by yunhaoling** [comment]
yes, our plan is to release in the middle of May, we'll update you once we get it shipped.

**Comment by yunhaoling** [comment]
Hey @engebret , sorry for keeping you waiting, we've released azure-servicebus v0.50.3 today, please try update the library accordingly.

One thing I want to note is that v0.50.3 would be our final release for v0.50.

Currently we're putting all our efforts into azure-servicebus v7.0.0 (which is in preview status) to create a more user friendly and productive library. As we have limited cycles, we decided not to maintain the old preview library which will never go GA.

Apologize for the inconvenience we bring to you and thanks for your understanding (and your collaboration on fixing this issue!).

We're looking forward to you trying our latest preview library and feedback is most welcome.
